pkg/controller/nginx/nginx.go

```diff
@@ -0,0 +1,109 @@
+package nginx
+
+import (
+	"github.com/Azure/aks-app-routing-operator/pkg/config"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/ingress"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/keyvault"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/osm"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/service"
+	"github.com/Azure/aks-app-routing-operator/pkg/manifests"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var (
+	defaultIngConfig = &manifests.NginxIngressConfig{
+		ControllerClass: "webapprouting.kubernetes.azure.com/nginx",
+		ResourceName:    "nginx",
+		IcName:          "webapprouting.kubernetes.azure.com",
+	}
+	ingConfigs = []*manifests.NginxIngressConfig{defaultIngConfig}
+)
+
+type nginx struct {
+	name       string
+	manager    manager.Manager
+	conf       *config.Config
+	self       *appsv1.Deployment
+	ingConfigs []*manifests.NginxIngressConfig
+}
+
+// New starts all resources required for Nginx ingresses
+func New(m manager.Manager, conf *config.Config, self *appsv1.Deployment) error {
+	n := &nginx{
+		name:       "nginx",
+		manager:    m,
+		conf:       conf,
+		self:       self,
+		ingConfigs: ingConfigs,
+	}
+
+	if err := n.addIngressClassReconciler(); err != nil {
+		return err
+	}
+
+	if err := n.addIngressControllerReconciler(); err != nil {
+		return err
+	}
+
+	if err := n.addConcurrencyWatchdog(); err != nil {
+		return err
+	}
+
+	if err := n.addIngressSecretProviderClassReconciler(); err != nil {
+		return err
+	}
+
+	if err := n.addPlaceholderPodController(); err != nil {
+		return err
+	}
+
+	if err := n.addIngressReconciler(); err != nil {
+		return err
+	}
+
+	if err := n.addIngressBackendReconciler(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (n *nginx) addIngressClassReconciler() error {
+	objs := []client.Object{}
+	for _, config := range n.ingConfigs {
+		objs = append(objs, manifests.NginxIngressClass(n.conf, n.self, config)...)
+	}
+
+	return ingress.NewIngressClassReconciler(n.manager, objs, n.name)
+}
+
+func (n *nginx) addIngressControllerReconciler() error {
+	objs := []client.Object{}
+	for _, config := range n.ingConfigs {
+		objs = append(objs, manifests.NginxIngressControllerResources(n.conf, n.self, config)...)
+	}
+
+	return ingress.NewIngressControllerReconciler(n.manager, objs, n.name)
+}
+
+func (n *nginx) addConcurrencyWatchdog() error {
+	return ingress.NewNginxConcurrencyWatchdog(n.manager, n.conf, n.ingConfigs)
```

Comment:
Here's what I'm seeing:

- the controller entrypoint (controller.go) constructs an `nginx` controller
- the nginx controller is implemented in large part by generic constructs in the `ingress` package
- in the future, other ingress controllers can use these same constructs

Is this accurate? If it is, we should use an interface type for the args receiving `n.ingConfigs` since `ingress`/`keyvault`/`osm` functions shouldn't be coupled to the nginx config. The interface can then expose getter functions that will be common across ingress implementations.

Reply:
Related: the keyvault, service, and osm controllers aren't related to nginx directly (i.e. they'll work with any ingress implementation). So they should be constructed in controller.go. The problem is getting the ingress configs to them, maybe a getter on the nginx controller would be alright?

Reply:
This is correct! I swapped to interfaces for the keyvault and osm functions. Will do the same for `NewNginxConcurrencyWatchdog` tomorrow (that's more work since it has some nginx specific things like metrics port and family.


A getter function on the nginx controller seems like the only way to get the ingress configs to them, unfortunately. I'd prefer that to passing them as an argument to the controller.

Reply:
Swapped all to generalized + moved the common ones back to the top-level controller